_stbt/core.py

```diff
@@ -1775,31 +1816,64 @@ def _find_match(image, template, match_parameters, imglog):
             else:
                 roi_mask = cv2.pyrUp(roi_mask)
 
-        matched, best_match_position, certainty, roi_mask = _match_template(
-            image_pyramid[level], template_pyramid[level], match_parameters,
+        heatmap = _match_template(
+            image_pyramid[level], template_pyramid[level], method,
             roi_mask, level, imglog)
 
+        # Relax the threshold slightly for scaled-down pyramid levels to
+        # compensate for scaling artifacts.
+        threshold = max(
+            0,
+            match_parameters.match_threshold - (0.2 if level > 0 else 0))
+
+        matched, best_match_position, certainty = _find_best_match_position(
+            heatmap, method, threshold, level)
+
         if not matched:
             break
 
-    # pylint: disable=W0631
-    return matched, _upsample(best_match_position, level), certainty
-
+        _, roi_mask = cv2.threshold(
+            heatmap,
+            ((1 - threshold) if method == cv2.TM_SQDIFF_NORMED else threshold),
+            255,
+            (cv2.THRESH_BINARY_INV if method == cv2.TM_SQDIFF_NORMED
+             else cv2.THRESH_BINARY))
+        roi_mask = roi_mask.astype(numpy.uint8)
+        imglog.add("source_matchtemplate_threshold", roi_mask, level)
 
-def _match_template(image, template, match_parameters, roi_mask, level, imglog):
+    while True:
+        # pylint:disable=undefined-loop-variable
+        yield (matched, _upsample(best_match_position, level), certainty)
+        if not matched:
+            return
+
+        # Exclude any positions that would overlap the previous match, then
+        # keep iterating until we don't find any more matches.
+        match_x, match_y = best_match_position
+        template_h, template_w = template.shape[:2]
+        region = Region(
+            x=match_x - (template_w - 1),
+            y=match_y - (template_h - 1),
+            right=match_x + template_w,
+            bottom=match_y + template_h)
+        mask_value = (255 if match_parameters.match_method == 'sqdiff-normed'
+                      else 0)
+        cv2.rectangle(
+            heatmap,
+            # -1 because cv2.rectangle considers the bottom-right point to be
+            # *inside* the rectangle.
+            (region.x, region.y), (region.right - 1, region.bottom - 1),
+            mask_value, cv2.cv.CV_FILLED)
+
+        matched, best_match_position, certainty = _find_best_match_position(
+            heatmap, method, threshold, level)
```

Comment:
Overall this seems fairly straight-forward.  My understanding is:

1. We split up the heat map generation from the finding the top location so we can do the former once and the latter many times
2. Then `_find_match -> _find_candidate_matches` that implements the first-stage matching becomes an iterator. This is the most complicated bit because of the Pyramid meaning that we might have to return/yield early.
3. `_match -> _find_matches` then iterates and checks each location from `_find_match` against the second pass, after checking that the first pass passed
4. `match` just returns the first result from `_find_matches`
5. `match_all` generates all but the last one